### ISSUE TITLE
Add option to include full path to file in application messages

### ISSUE THREAD

**Issue by habamax** [asciidoctor/asciidoctor#3414]
Imagine I have a document `~/docs/test/test.adoc`.

My current directory is `~/docs` and I issue the `asciidoctor-pdf` command:
```
asciidoctor-pdf "~/docs/test/test.adoc"
```

If there is an error in a file I get a message like:

```
asciidoctor: ERROR: test.adoc: line 20: level 0 sections can only be used when doctype is book
```

File `test.adoc` in the message is without any path specified which makes it impossible to jump to the error from within vim. (It just goes to new `~/docs/test.adoc` file instead).

Is it possible to tell asciidoctor to print full or relative to the directory where asciidoctor was executed file/path?

PS, I tried to specify `--source-dir "~/docs"` but it doesn't change anything.

PPS, vim errorformat I try to use:

`setlocal errorformat=asciidoctor:\ ERROR:\ %f:\ line\ %l:\ %m`

and it looks in vim:

![image](https://user-images.githubusercontent.com/234774/64798246-5da86c00-d58b-11e9-97d3-9ffa26cd84b3.png)

**Comment by hupfdule** [comment]
An alternative that could help it issuing a line telling that the asciidoc processor switches to a given directory before executing like c compilers do.

Something like this:

```
$ pwd
/home/myhome
$ asciidoctor /tmp/somefile.adoc
Changing to directory: /tmp/
asciidoctor: WARNING: somefile.adoc: line 83: unterminated open block
asciidoctor: WARNING: chapter1/included.adoc: line 7: section title out of sequence: expected level 2, got level 4

```

**Comment by habamax** [comment]
> An alternative that could help it issuing a line telling that the asciidoc processor switches to a given directory before executing like c compilers do.

I think it is easier/simpler to provide proper file name in errors/warnings

**Comment by mojavelinux** [comment]
The reason the current message is how it is, historically, is to be consistent with AsciiDoc Python. It also prevents the message from wrapping excessively. However, I recognize in large systems, this drops very important context. So I'm open to being able to include more detail in the error message. I'm thinking that this could be some sort of compatibility setting. We couldn't just change it outright by default because existing applications may be relying on the output as it is. So that would be a major breaking change. But the compatibility setting could give us the best of both worlds.

My main concern about this issue is that the subject is hiding the real change. Although this may be useful in VIM, it has nothing at all to do with VIM. So I'm going to update the subject so it better reflects the change being requested.

**Comment by hupfdule** [comment]
Actually the path doesn't need to be the full path (which would actually be annoying in most cases).

The main problem is that asciidoctor doesn't take difference between the _current directory_ and _the directory containing the asciidoc file_ into account.

For example here:

```
$ pwd
/home/myhome
$ asciidoctor documents/project1/somefile.adoc
asciidoctor: WARNING: somefile.adoc: line 83: unterminated open block
asciidoctor: WARNING: chapter1/included.adoc: line 7: section title out of sequence: expected level 2, got level 4
 ```

the information about *documents/project1* is lost. And that should be included then, because the given file names would be acessible with that path from the current working directory:

```
$ pwd
/home/myhome
$ asciidoctor documents/project1/somefile.adoc
asciidoctor: WARNING: documents/project1/somefile.adoc: line 83: unterminated open block
asciidoctor: WARNING: documents/project1/chapter1/included.adoc: line 7: section title out of sequence: expected level 2, got level 4
 ```

**Comment by habamax** [comment]
What seemed to be an easy thing at first, became a quite an issue :)

**Comment by habamax** [comment]
> Actually the path doesn't need to be the full path (which would actually be annoying in most cases).

For my purpose it actually doesn't matter if it is a full path or relative to the current working dir. Anyway it is considered as "breaking" change so default behavior should stay the same. And probably we can have an option to either provide a full or relative path to a file.

**Comment by hupfdule** [comment]
BTW there already is an incompatibilty between asciidoc and asciidoctor regarding the reported errors:

For this example document:

```
= My Title

some text

== Some section

some more text

// The offending line:
----

and even more text
```

the output is as follows:

```
asciidoc: ERROR: file.adoc: line 9: [blockdef-listing] missing closing delimiter
```

```
asciidoctor: WARNING: file.adoc: line 10: unterminated listing block
```

1. `asciidoc` issues an ERROR where `asciidoctor` only issues a WARNING
   I think an error makes more sense here, because usually the rendering of the remaining file is broken.
2. `asciidoc` issues the _before_ the actual problem (only in this case, others are correct) whereas `asciidoctor` issues the actual problematic line
   Of course `asciidoctor` is correct here, but still it is different from `asciidoc`

**Comment by mojavelinux** [comment]
> BTW there already is an incompatibilty between asciidoc and asciidoctor regarding the reported errors:

I didn't mean to suggest that the messages are exactly the same. They are definitely not. They have diverged over time. I was explaining the reason the *path* is reported as it is was to be compatible with AsciiDoc Python.

**Comment by mojavelinux** [comment]
> The main problem is that asciidoctor doesn't take difference between the current directory and the directory containing the asciidoc file into account.

That's correct, but also intentional. Asciidoctor is reporting the path from the base_dir (or docdir of the primary document, if base_dir is not specified).

**Comment by mojavelinux** [comment]
> And probably we can have an option to either provide a full or relative path to a file.

I definitely support exploring the idea of making this an option.

**Comment by habamax** [comment]
> I definitely support exploring the idea of making this an option.

How are options handled in asciidoctor? Could you point me where it is processed (found it `asciidoctor/lib/asciidoctor/cli/options.rb`) and how do you access them?

**Comment by mojavelinux** [comment]
Options are passed via the API. There are options supported on the CLI, but those are selective (because not all can be expressed that way).

See https://asciidoctor.org/docs/user-manual/#ruby-api-options

However, in this case, it may need to be a compliance option since the logger is global. A compliance option is also set via the API, but differently. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor.rb#L121-L175

I won't know which one will work until I explore it some.
